[SPARK-11977][SQL] Support accessing a column contains "." without backticks

rerngvit opened the pull request:
## What changes were proposed in this pull request?
- Add support for accessing a dataframe column that contains "." in its name without backticks
- Add a testcase for this in DataFrameSuite
## How was this patch tested?

Spark SQL unit test and manual testing

rerngvit commented:
@sun-rui

AmplabJenkins commented:
Can one of the admins verify this patch?

rerngvit commented:
@shivaram

jaceklaskowski commented on sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:
```diff
@@ -641,6 +641,10 @@ class DataFrameSuite extends QueryTest with SharedSQLContext {
+
+    // Renaming to a column that contains "." character
+    val df2 = testData.toDF().withColumnRenamed("value", "value.Renamed")
+    assert(df2.schema.map(_.name) === Seq("key", "value.Renamed"))
```
`df2.schema.fieldNames`?

sun-rui replied:
Please add more test cases that columns whose name has '.' can be accessed without backticks

jaceklaskowski commented on sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:
```diff
@@ -641,6 +641,10 @@ class DataFrameSuite extends QueryTest with SharedSQLContext {
     assert(df.schema.map(_.name) === Seq("key", "valueRenamed", "newCol"))
+
+    // Renaming to a column that contains "." character
+    val df2 = testData.toDF().withColumnRenamed("value", "value.Renamed")
```
No need for `()` in `toDF()`.

jaceklaskowski commented on sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:
```diff
@@ -641,6 +641,10 @@ class DataFrameSuite extends QueryTest with SharedSQLContext {
         Row(key, value, key + 1)
       }.toSeq)
     assert(df.schema.map(_.name) === Seq("key", "valueRenamed", "newCol"))
```
I'd fix that line, too with `fieldNames`

rerngvit replied:
@jaceklaskowski: I agree with you that the code can be made shorter and more elegant in the ways that you pointed out. Nonetheless, the improvements: (1) removing () in toDF() and (2) replacing the schema map with fieldNames, are actually applicable to other places in the file "DataFrameSuite.scala" as well. In other words, it is not directly related to this PR. I think the proper way to handle this is rather in another JIRA issue. Feel free to fire it.

shivaram commented:
I'm not sure I am good reviewer for this as I dont fully understand the consequences inside SQL for this change. cc @liancheng @rxin

Commit:
[SPARK-11977][SQL] Support accessing a DataFrame column using its name without backticks if the name contains '.'

## What changes were proposed in this pull request?
- Add support for accessing dataframe column that contains "." in its name without backticks
- Add testcases for this in DataFrameSuite

## How was this patch tested?
Spark unit tests and manual testing

rerngvit commented:
@sun-rui I added more testcases for this. Please have a look. It would be great, if you could enable the Jenkins CI testing. I don't have a permission to do so.

sun-rui commented:
Does this solution co-operate with the access pattern of "table.column"?

rerngvit commented:
Yes.

shivaram commented:
Jenkins, ok to test

SparkQA commented:
**[Test build #62831 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62831/consoleFull)** for PR 14309 at commit [`0e47cfd`](https://github.com/apache/spark/commit/0e47cfd47774a8d75e9b45bbeb5a19fddb40494e).

felixcheung commented on sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:
```diff
@@ -896,6 +896,19 @@ class DataFrameSuite extends QueryTest with SharedSQLContext {
   }
 
+  test("SPARK-11977: Support accessing a column contains '.' without backticks") {
+    val df = sparkContext.parallelize(
```
could you add some test cases with a DataFrame created from `iris`?

SparkQA commented:
**[Test build #62831 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62831/consoleFull)** for PR 14309 at commit [`0e47cfd`](https://github.com/apache/spark/commit/0e47cfd47774a8d75e9b45bbeb5a19fddb40494e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62831/
Test PASSed.

sun-rui commented:
@cloud-fan, could you help to review it?

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:
```diff
@@ -896,6 +896,19 @@ class DataFrameSuite extends QueryTest with SharedSQLContext {
+  test("SPARK-11977: Support accessing a column contains '.' without backticks") {
+    val df = sparkContext.parallelize(
+      (1, 2) :: (3, 4) ::  Nil).toDF("test.column1", "test.column.2")
+    checkAnswer(df.select("test.column1"), Seq(Row(1), Row(3)))
```
This is a behaviour change, users may expect to see an analyse failure here. cc @rxin Do we want this feature?

Personally I'd like to have a new function(similar to `col`, but resolves the column by the given name directly) instead of changing the semantic of `select`

rerngvit replied:
Any follow up on this?

rerngvit commented:
I have not got any replies on this. I presume that this is a soft reject for this PR.
